test/TypeCheck.lean

```diff
@@ -0,0 +1,11 @@
+import Mathlib.Tactic.TypeCheck
+
+example : True := by {
+  type_check Nat; -- Type
+  type_check Bool.true; -- Bool
+  type_check nat_lit 1; -- Nat
+  type_check (1 : Nat); -- Nat
+  type_check (True : _); -- Prop
+  type_check ∀ x y : Nat, x = y; -- Prop
+  type_check fun x : Nat => 2 * x + 1; -- Nat -> Nat
+  exact True.intro }
```

Comment:
```suggestion
example : True := by
  type_check Nat -- Type
  type_check Bool.true -- Bool
  type_check nat_lit 1 -- Nat
  type_check (1 : Nat) -- Nat
  type_check (True : _) -- Prop
  type_check ∀ x y : Nat, x = y -- Prop
  type_check fun x : Nat => 2 * x + 1 -- Nat -> Nat
  trivial
```

Reply:
This needs some type check error tests as well, using `fail_if_success`.

Reply:
Marking as unresolved, as we still need to error tests.

Reply:
I'm not sure I understand, can you explain in more detail what we need?

Reply:
e.g. `fail_if_success type_check 2 + "hi"`

Reply:
#334 has an alternative `expect_failure` tactic which should give you what you want here. We'll try to merge that soon?

Reply:
Oh dear, `type_check 2 + "hi"` fails correctly with `failed to synthesize instance HAdd Nat String ?m.160`, but `expect_failure type_check 2 + "hi"` (after merging master into this PR) claims `tactic sequence succeeded`.

Reply:
This is a bug in `fail_if_success`/`success_if_fail`/`expect_failure` (the synonyms keep piling up...), not `type_check`.

Reply:
> could you suggest how we should be testing this?

I'm not 100% sure what `type_check` is supposed to do, but I guess it's supposed to find bugs in proofs produced by tactics.  So I would suggest to test a tactic or elaborator that produces a term which does not type check.  Such as this one:

```lean
/-- A term where `inferType` returns `Prop`, but which does not type check. -/
elab "wrong" : term =>
  return Lean.mkApp2 (.const ``id [.zero]) (.sort .zero) (.app (.sort .zero) (.sort .zero))
```

I have committed this elaborator and a failing test case.